modules/core/04-channel/keeper/handshake.go

```diff
@@ -394,43 +370,38 @@ func (k Keeper) ChanOpenConfirm(
 		)
 	}
 
-	// verify multihop proof or standard proof
+	var counterpartyHops []string
 	if len(channel.ConnectionHops) > 1 {
-		kvGenerator := func(mProof *types.MsgMultihopProofs, _ *connectiontypes.ConnectionEnd) (string, []byte, error) {
-			key := host.ChannelPath(channel.Counterparty.PortId, channel.Counterparty.ChannelId)
-
-			counterpartyHops, err := mProof.GetCounterpartyConnectionHops(k.cdc, &connectionEnd)
-			if err != nil {
-				return "", nil, err
-			}
-			counterparty := types.NewCounterparty(portID, channelID)
-			expectedChannel := types.NewChannel(
-				types.OPEN, channel.Ordering, counterparty,
-				counterpartyHops, channel.Version,
-			)
-			value, err := expectedChannel.Marshal()
-			if err != nil {
-				return "", nil, err
-			}
-			return key, value, nil
+		var (
+			err           error
+			multihopProof types.MsgMultihopProofs
+		)
+		if err = k.cdc.Unmarshal(proofAck, &multihopProof); err != nil {
+			return err
 		}
 
-		return k.connectionKeeper.VerifyMultihopMembership(
-			ctx, connectionEnd, proofHeight, proofAck,
-			channel.ConnectionHops, kvGenerator)
-
+		counterpartyHops, err = multihopProof.GetCounterpartyConnectionHops(k.cdc, &connectionEnd)
+		if err != nil {
+			return err
+		}
 	} else {
-		counterpartyHops := []string{connectionEnd.GetCounterparty().GetConnectionID()}
-		counterparty := types.NewCounterparty(portID, channelID)
-		expectedChannel := types.NewChannel(
-			types.OPEN, channel.Ordering, counterparty,
-			counterpartyHops, channel.Version)
-
-		return k.connectionKeeper.VerifyChannelState(
-			ctx, connectionEnd, proofHeight, proofAck,
-			channel.Counterparty.PortId, channel.Counterparty.ChannelId,
-			expectedChannel)
+		counterpartyHops = []string{connectionEnd.GetCounterparty().GetConnectionID()}
```

Comment:
This code is pretty much repeated in all handlers, so we could rug it under some common function.